### PR TITLE
Allow templating of email subjects

### PR DESCRIPTION
**Pull request by nickvergessen**

### CODE REVIEW

**Review by MorrisJobke — CHANGES_REQUESTED** [review]
Unit tests fail

**Review by LukasReschke — DISMISSED** [review]
Autoloader is not up-to-date

**Comment by nickvergessen** [comment]
Rebased, please review

File: lib/private/Mail/Mailer.php [hunk 1]
```diff
@@ -124,13 +125,13 @@ public function createEMailTemplate($emailId, array $data = []) {
 	 * Send the specified message. Also sets the from address to the value defined in config.php
 	 * if no-one has been passed.
 	 *
-	 * @param Message $message Message to send
+	 * @param IMessage|Message $message Message to send
```
Comment (MorrisJobke) [inline lib/private/Mail/Mailer.php]: `IMessage` is fully enough here.

File: lib/public/Mail/IMessage.php [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+	 * @since 13.0.0
+	 */
+	public function setBcc(array $recipients);
+
+	/**
+	 * @param IEMailTemplate $emailTemplate
+	 * @return $this
+	 * @since 13.0.0
+	 */
+	public function setTemplate(IEMailTemplate $emailTemplate);
```
Comment (MorrisJobke) [inline lib/public/Mail/IMessage.php]: I would call it `useTemplate()` just to make it more clear why the template is passed in - to use it as base for the subject, html and plain body. And also write this in the description of the interface.

**Review by MorrisJobke — CHANGES_REQUESTED** [review]

**Comment by MorrisJobke** [comment]
Beside that it works nice 👍

File: lib/private/Mail/Mailer.php (hunk 1 above)
Comment (nickvergessen) [reply lib/private/Mail/Mailer.php]: No, because we use methods of the non public api in here ;)

File: lib/public/Mail/IMessage.php (hunk 2 above)
Comment (nickvergessen) [reply lib/public/Mail/IMessage.php]: I thought the same, but it only extracts them once, so when you change them later they are not updated anymore.
But I think set is also misleading there, suggestions welcome

Comment (MorrisJobke) [reply lib/public/Mail/IMessage.php]: Then still `useTemplate()` makes most sense and is fine - even if it is used only once.

Comment (nickvergessen) [reply lib/public/Mail/IMessage.php]: Done and rebased

**Review by MorrisJobke — APPROVED** [review]
Tested and works 👍

**Review by rullzer — APPROVED** [review]

**Pull request merged**